src/rgw/rgw_amqp_1.cc

```diff
@@ -0,0 +1,542 @@
+      case proton::transfer::state::RECEIVED:
+        return "AMQP_1_TRACKER_STATUS_RECEIVED";
+      case proton::transfer::state::ACCEPTED:
+        return "AMQP_1_TRACKER_STATUS_ACCEPTED";
+      case proton::transfer::state::REJECTED:
+        return "AMQP_1_TRACKER_STATUS_REJECTED";
+      case proton::transfer::state::RELEASED:
+        return "AMQP_1_TRACKER_STATUS_RELEASED";
+      case proton::transfer::state::MODIFIED:
+        return "AMQP_1_TRACKER_STATUS_MODIFIED";
+    }
+    return "";
+  }
+
+  int reply_to_code(enum proton::transfer::state s) {
+    switch(s) {
+      case proton::transfer::state::NONE:
+        return AMQP_1_TRACKER_STATUS_UNKOWN;
+      case proton::transfer::state::RECEIVED:
+        return AMQP_1_TRACKER_STATUS_RECEIVED;
+      case proton::transfer::state::ACCEPTED:
+        return AMQP_1_TRACKER_STATUS_ACCEPTED;
+      case proton::transfer::state::REJECTED:
+        return AMQP_1_TRACKER_STATUS_REJECTED;
+      case proton::transfer::state::RELEASED:
+        return AMQP_1_TRACKER_STATUS_RELEASED;
+      case proton::transfer::state::MODIFIED:
+        return AMQP_1_TRACKER_STATUS_MODIFIED;
+    }
+    return AMQP_1_TRACKER_STATUS_UNKOWN;
+  }
+
+  class connection_t : public proton::messaging_handler{
+    friend class Manager;
+    public:
+    bool marked_for_deletion = false;
+    bool ready = false;
+    uint64_t delivery_tag = 1;
+    int status;
+    mutable std::atomic<int> ref_count = 0;
+    CephContext* const cct;
+    MessageQueue messages;
+    CallbackList callbacks;
+    std::string broker;
+    proton::connection connection;
+    proton::sender sender;
+    // proton::connection_options options;
+    proton::work_queue* pwork_queue;
+    int queued;
+
+    const boost::optional<std::string> ca_location;
+
+    public:
+
+    void on_connection_open(proton::connection& conn) {
+      connection = conn;
+      ldout(cct, 10) << "AMQP 1.0 proton: connnection opened." << dendl;
+    }
+
+    void on_connection_close(proton::connection& conn) {
+      ldout(cct, 10) << "AMQP 1.0 proton: connection closed." << dendl;
+    }
+
+    void on_connection_error(proton::connection& conn) {
+      ldout(cct, 1) << "AMQP 1.0 proton: connection error:" << conn.error().what() << dendl;
+    }
+
+    // assign the work_queue of the sender
+    void on_sender_open(proton::sender& s) override {
+      sender = s;
+      pwork_queue = &s.work_queue();
+      ready = true;
+    }
+
+    void on_sender_error(proton::sender& s) override {
+      status = RGW_AMQP_1_STATUS_SENDER_ERROR;
+    }
+
+    void on_sendable(proton::sender& s) override {
+      auto count = messages.consume_all(std::bind(&connection_t::send, this,
+        std::placeholders::_1));
+      queued -= count;
+    }
+
+    // send a message, and keep record the tracker for later callbacks
+    void send(message_wrapper_t* message) {
+      // t is the tracker of this message
+      pwork_queue->add([=]() {
```

Comment:
please capture only things that you actually need in the lambda. doing "=" capture is usually not needed